packages/runner-ct/src/iframe/iframes.tsx

```diff
@@ -22,134 +21,133 @@ export function getSpecUrl ({ namespace, spec }, prefix = '') {
+
+  const [iframeModel] = useState(
+    new IframeModel({
+      state: props.state,
+      restoreDom: autIframe.restoreDom,
+      highlightEl: autIframe.highlightEl,
+      detachDom: autIframe.detachDom,
+      snapshotControls: (snapshotProps) => (
+        <SnapshotControls
+          eventManager={props.eventManager}
+          snapshotProps={snapshotProps}
+          state={props.state}
+          onToggleHighlights={_toggleSnapshotHighlights}
+          onStateChange={_changeSnapshotState}
         />
-        <ScriptError error={scriptError} />
-        <div className='cover' />
-      </div>
-    )
-  }
-
-  componentDidMount () {
-    const config = this.props.config
-
-    this.autIframe = new AutIframe(config)
+      ),
+    }),
+  )
 
-    this.props.eventManager.on('visit:failed', this.autIframe.showVisitFailure)
-    this.props.eventManager.on('before:screenshot', this.autIframe.beforeScreenshot)
-    this.props.eventManager.on('after:screenshot', this.autIframe.afterScreenshot)
-    this.props.eventManager.on('script:error', this._setScriptError)
+  useEffect(() => {
+    props.eventManager.on('visit:failed', autIframe.showVisitFailure)
+    props.eventManager.on('before:screenshot', autIframe.beforeScreenshot)
+    props.eventManager.on('after:screenshot', autIframe.afterScreenshot)
+    props.eventManager.on('script:error', _setScriptError)
 
     // TODO: need to take headless mode into account
     // may need to not display reporter if more than 200 tests
-    this.props.eventManager.on('restart', () => {
-      this._run(this.props.state.spec, config)
+    props.eventManager.on('restart', () => {
+      _run(props.state.spec, props.config)
     })
 
-    this.props.eventManager.on('print:selector:elements:to:console', this._printSelectorElementsToConsole)
+    props.eventManager.on('print:selector:elements:to:console', autIframe.printSelectorElementsToConsole)
+
+    props.eventManager.start(props.config)
+
+    setDisposers([
+      autorun(() => {
+        autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
+      }),
+      autorun(() => {
+        autIframe.toggleSelectorHighlight(selectorPlaygroundModel.isShowingHighlight)
+      }),
+      autorun(() => {
+        if (props.state.spec) {
+          _run(props.state.spec, props.config)
+        }
+      }),
+    ])
+
+    iframeModel.listen()
+
+    return () => {
+      props.eventManager.notifyRunningSpec(null)
+      props.eventManager.stop()
+      disposers.forEach((dispose) => {
+        dispose()
+      })
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
 
-    this._disposers.push(autorun(() => {
-      this.autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
-    }))
+  useEffect(() => {
+    props.state.callbackAfterUpdate?.()
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [props.state.callbackAfterUpdate])
```

Comment:
I don't know why it's complaining here.

Also, I think I'm supposed to pass this as a separate prop that is wrapped in a `useCallback`. Is this correct? @agg23

Reply:
It's complaining because it doesn't like the deep property access. It would stop if you assigned `props.state.callbackAfterUpdate` to a variable.

I don't understand your question.

Reply:
I see - this warning is basically a false positive, then. This pattern (deep property access) is used in the React docs:

![image](https://user-images.githubusercontent.com/19196536/119283109-594cb580-bc7f-11eb-88d3-141890c26911.png)